core/src/components/picker-internal/picker-internal.tsx

```diff
@@ -0,0 +1,504 @@
+      this.destroyKeypressListener = () => {
+        el.removeEventListener('keypress', this.onKeyPress);
+      }
+    }
+
+    this.emitInputModeChange();
+  }
+
+  private exitInputMode = () => {
+    const { inputEl, inputMode } = this;
+    if (!inputMode || !inputEl) { return; }
+
+    this.inputMode = false;
+    this.inputModeColumn = undefined;
+    inputEl.blur();
+    inputEl.value = '';
+
+    if (this.destroyKeypressListener) {
+      this.destroyKeypressListener();
+      this.destroyKeypressListener = undefined;
+    }
+
+    this.emitInputModeChange();
+  }
+
+  private onKeyPress = (ev: KeyboardEvent) => {
+    const { inputEl } = this;
+    if (!inputEl) { return; }
+
+    const parsedValue = parseInt(ev.key, 10);
+
+    /**
+     * Only numbers should be allowed
+     */
+    if (!Number.isNaN(parsedValue)) {
+      inputEl.value += ev.key;
+
+      this.onInputChange()
+    }
+  }
+
+  private selectSingleColumn = () => {
+    const { inputEl, inputModeColumn, singleColumnSearchTimeout } = this;
+    if (!inputEl || !inputModeColumn) { return; }
+
+    const values = inputModeColumn.items;
+    let valueToSelect;
+
+    /**
+     * If users pause for a bit, the search
+     * value should be reset similar to how a
+     * <select> behaves. So typing "34", waiting,
+     * then typing "5" should select "05".
+     */
+    if (singleColumnSearchTimeout) {
+      clearTimeout(singleColumnSearchTimeout);
+    }
+
+    this.singleColumnSearchTimeout = setTimeout(() => {
+      inputEl.value = '';
+      this.singleColumnSearchTimeout = undefined;
+    }, 1000);
+
+    /**
+     * For values that are longer than 2 digits long
+     * we should shift the value over 1 character
+     * to the left. So typing "456" would result in "56".
+     */
+    if (inputEl.value.length >= 3) {
+      const startIndex = inputEl.value.length - 2;
+      const newString = inputEl.value.substring(startIndex);
+
+      inputEl.value = newString;
+      this.selectSingleColumn();
+      return;
+    }
+
+    /**
+     * Checking the value of the input gets priority
+     * first. For example, if the value of the input
+     * is "1" and we entered "2", then the complete value
+     * is "12" and we should select hour 12.
+     */
+    const findItemFromCompleteValue = values.find(v => v.text === `0${inputEl.value}` || v.text === inputEl.value);
+    if (findItemFromCompleteValue) {
+      inputModeColumn.value = findItemFromCompleteValue.value;
+      return;
+    }
+
+    /**
+     * If we typed "56" to get minute 56, then typed "7",
+     * we should select "07" as "567" is not a valid minute.
+     */
+    if (inputEl.value.length === 2) {
+      const changedCharacter = inputEl.value.substring(inputEl.value.length - 1);
+      inputEl.value = changedCharacter;
+      this.selectSingleColumn();
+    }
+  }
+
+  private searchHourColumn = (colEl: HTMLIonPickerColumnInternalElement, value: string) => {
```

Comment:
I would merge this and `searchMinuteColumn`; maybe prioritize checking for matches with leading 0s, and if none are found, check again with trailing 0s? Or something similar but more efficient than that, haha.